### PR TITLE
Added support for query parameters for older versions, plus bug fix

### PR DESCRIPTION
**Pull request by Treno1**
Started as fix for #28 , but then understood that most of query parameters types doesn`t work.
For example AddParameter("var", new [] {1,2,3}); will result in ClickHouseException cause int array ToString() will result in "Int32[]", but not the data.

Checked on yandex/clickhouse-server:19.9 version. Works fine with known types (Enum and DateTime64 aren`t supported in this version)

### CODE REVIEW

File: ClickHouse.Client/Utility/CommandExtensions.cs [hunk 1]
```diff
@@ -2,17 +2,27 @@
...
 using System.Threading.Tasks;
 using ClickHouse.Client.ADO.Adapters;
+using ClickHouse.Client.ADO.Parameters;
 
 namespace ClickHouse.Client.Utility
 {
     public static class CommandExtensions
     {
-        public static DbParameter AddParameter(this DbCommand command, string parameterName, object parameterValue)
+        public static ClickHouseDbParameter AddParameter(this DbCommand command, string parameterName, object parameterValue)
```
Comment (DarkWanderer) [inline ClickHouse.Client/Utility/CommandExtensions.cs:11]: If this extension returns `ClickHouseDbParameter`, it should be applicable only to `ClickHouseCommand`

File: ClickHouse.Client/Types/PlainDataType.cs [hunk 2]
```diff
@@ -17,5 +23,60 @@ public PlainDataType(ClickHouseTypeCode dataType)
         public override ClickHouseTypeCode TypeCode => dataType;
 
         public override string ToString() => TypeCode.ToString();
+        
+        public override string ToHttpParameter(object value)
```
Comment (DarkWanderer) [inline ClickHouse.Client/Types/PlainDataType.cs:27]: This looks like it should not be part of type system, but a separate class like 'HttpFormatter' in `Formats`

File: ClickHouse.Client/Types/ClickHouseType.cs [hunk 3]
```diff
@@ -8,6 +8,12 @@ internal abstract class ClickHouseType
 
         public abstract Type FrameworkType { get; }
 
+        public abstract string ToHttpParameter(object value);
```
Comment (DarkWanderer) [inline ClickHouse.Client/Types/ClickHouseType.cs:11]: This should really be in separate class (like `BinaryStreamWriter`, but for http)

File: ClickHouse.Client.Tests/SqlSelectTests.cs [hunk 4]
```diff
@@ -204,21 +204,53 @@ public async Task ShouldGetReaderSchemaTable()
...
         }
+        
+        public static IEnumerable<TestCaseData> ParametersQueries => TestUtilities.GetDataTypeSamples()
+            //.Where(sample => !new [] {"Enum", "DateTime64(9)"}.Contains(sample.ClickHouseType)) //old clh doesn`t know about regular Enum and DateTime64
+            .Where(sample => sample.ExampleValue != DBNull.Value) //null value should be handled by writing "is null" statement
+            .Where(sample => sample.ClickHouseType != "Date") //DateTime with 00:00:00 can`t be Date type, cause it`ll broke DateTime logic. Specify type to fix that
+            .Where(sample => sample.ClickHouseType != "DateTime64(9)") //Default DateTime CLH analog is DateTime. If you want to use DateTime64 specify type as parameter
+            .Where(sample => sample.ClickHouseType != "UUID") // https://github.com/ClickHouse/ClickHouse/issues/7463
+            .Select(sample => new TestCaseData(
+                $"SELECT * FROM (SELECT {sample.ExampleExpression} AS res) WHERE {sample.WhereClause}",
```
Comment (DarkWanderer) [inline ClickHouse.Client.Tests/SqlSelectTests.cs:215]: Testing 2 things at once, not a good idea (hard to understand what to fix if something breaks). `SELECT` and `WHERE` should be tested separately

File: ClickHouse.Client.Tests/TestUtilities.cs [hunk 5]
```diff
@@ -36,63 +36,66 @@ public struct DataTypeSample
...
 
         public static IEnumerable<DataTypeSample> GetDataTypeSamples()
         {
-            yield return new DataTypeSample("Nothing", typeof(DBNull), "NULL", DBNull.Value);
+            yield return new DataTypeSample("Nothing", typeof(DBNull), "NULL", DBNull.Value, "res is {var:Nothing}");
 
-            yield return new DataTypeSample("Int8", typeof(sbyte), "toInt8(-8)", -8);
-            yield return new DataTypeSample("UInt8", typeof(byte), "toUInt8(8)", 8);
+            yield return new DataTypeSample("Int8", typeof(sbyte), "toInt8(-8)", -8, "res = {var:Int8}");
+            yield return new DataTypeSample("UInt8", typeof(byte), "toUInt8(8)", 8, "res = {var:UInt8}");
```
[the remark this answers is not in the document]
Comment (Treno1) [reply ClickHouse.Client.Tests/TestUtilities.cs]: done

File: ClickHouse.Client/Utility/BackwardCompabilityExtension.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+                if (currentVersion[i] > targetVersion[i])
+                    return true;
+            }
+
+            return true;
+        }
+
+        private static int[] ParseVersion(string version)
+        {
+            return version.Split('.').Select(int.Parse).ToArray();
```
[the remark this answers is not in the document]
Comment (Treno1) [reply ClickHouse.Client/Utility/BackwardCompabilityExtension.cs]: done

File: ClickHouse.Client/Utility/CommandExtensions.cs (hunk 1 above)
Comment (Treno1) [reply ClickHouse.Client/Utility/CommandExtensions.cs]: done

File: ClickHouse.Client.Tests/SqlSelectTests.cs (hunk 4 above)
Comment (Treno1) [reply ClickHouse.Client.Tests/SqlSelectTests.cs]: done

File: ClickHouse.Client/Types/PlainDataType.cs (hunk 2 above)
Comment (Treno1) [reply ClickHouse.Client/Types/PlainDataType.cs]: done

File: ClickHouse.Client/Types/ClickHouseType.cs (hunk 3 above)
Comment (Treno1) [reply ClickHouse.Client/Types/ClickHouseType.cs]: done

**Comment by Treno1** [comment]
Fixed build. I originally worked from osx with .net core 3, but .net framework 4.7.2 doesn`t know about Dictionary.TryAdd

**Comment by DarkWanderer** [comment]
Nice! Thank you, merged

**Pull request merged**